Docs: Lifecycle Management of Counters and Lists in REST (#3560)
googleforgames/agones: site/content/en/docs/Guides/Client SDKs/rest.md

curl -d '{"key": "foo", "value": "bar"}' -H "Content-Type: application/json" -X PUT http://localhost:${AGONES_SDK_HTTP_PORT}/metadata/annotation
```

### Counters

> {{< alpha title="Counters" gate="Counters" >}}

Reviewer: ```suggestion
{{< alpha title="Counters and Lists" gate="CountsAndLists" >}}
```

There is only one feature flag for both Counters and Lists - so it needs to be both - hence I left in the top level header of "Counters and Lists" over each "Counters" and "Lists" heading.

Author: Included the header.

> **Note:** In the Counters examples, we retrieve the counter under the key "conformanceTestCounter" to make it extra clear.

Reviewer: ```suggestion
In all the  Counter examples, we retrieve the counter under the key `rooms` as if it was previously defined in [`GameServer.Spec.counters[room]`]({{< ref "/docs/reference/agones_crd_api_reference/#agones.dev/v1.GameServerSpec" >}})

For your own Counter REST requests, replace the value `rooms` with your own key in the path.
```

Author: done!

#### Alpha: GetCounter
This function retrieves a specified counter by its name and returns its information.

##### Example

```bash

> curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/conformanceTestCounter

Reviewer: ```suggestion
curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/rooms
```

(and all the other references)

Author: Renamed to `rooms` in all the references

```

Response:
```json
{"name":"conformanceTestCounter", "count":"1", "capacity":"10"}
```

#### Alpha: UpdateCounter
This function updates the specified counter's properties, such as its count and capacity, and returns the updated counter details.

##### Example

```bash
curl -d '{"count": "5", "capacity": "11"}' -H "Content-Type: application/json" -X PATCH http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/conformanceTestCounter
```

Response:
```json
{"name":"conformanceTestCounter", "count":"5", "capacity":"11"}
```

### Lists

{{< alpha title="Lists" gate="Lists" >}}

> **Note:** In the Lists examples, we retrieve the List under the key "conformanceTestList" to make it extra clear.

Reviewer: ```suggestion
In all the List examples, we retrieve the list under the key `players` as if it was previously defined in [`GameServer.Spec.lists[players]`]({{< ref "/docs/reference/agones_crd_api_reference/#agones.dev/v1.GameServerSpec" >}})

For your own List REST based requests, replace the value `players` with your own key in the path.
```

Author: changed it as per your suggestion

#### Alpha: GetList
This function retrieves a specific list identified by its name, returns the list's information.

##### Example
```bash
curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/lists/conformanceTestList
```

Response:
```json

> {"name":"conformanceTestList", "capacity":"100", "values":["test0", "test1", "test2"]}     

Reviewer: ```suggestion
{"name":"players", "capacity":"100", "values":["player0", "player1", "player2"]}     
```

And same as other `conformanceTestList` references below.

Author: Renamed to `players` in all the references.

curl -d '{"key": "foo", "value": "bar"}' -H "Content-Type: application/json" -X PUT http://localhost:${AGONES_SDK_HTTP_PORT}/metadata/annotation
```

### Counters and Lists

{{< alpha title="Counters and Lists" gate="CountsAndLists" >}}

#### Counters

In all the  Counter examples, we retrieve the counter under the key `rooms` as if it was previously defined in [`GameServer.Spec.counters[room]`]({{< ref "/docs/Reference/agones_crd_api_reference.html#agones.dev/v1.GameServerSpec" >}}).

For your own Counter REST requests, replace the value `rooms` with your own key in the path.

##### Alpha: GetCounter

> This function retrieves a specified counter by its name and returns its information.

Author: I've made the suggested changes. Ivy, wdyt?

###### Example

```bash
curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/rooms
```

Response:
```json
{"name":"rooms", "count":"1", "capacity":"10"}
```

##### Alpha: UpdateCounter

> This function updates the specified counter's properties, such as its count and capacity, and returns the updated counter details.

Author: Would it be a good idea if we rephrase like this?
This function updates the properties of the counter with the key `rooms`, such as its count and capacity, and returns the updated counter details.

> This function updates the specified counter's properties, such as its count and capacity, and returns the updated counter details.

Reviewer: That reads better! Nice.

###### Example

```bash
curl -d '{"count": "5", "capacity": "11"}' -H "Content-Type: application/json" -X PATCH http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/rooms
```

Response:
```json
{"name":"rooms", "count":"5", "capacity":"11"}
```

#### Lists

In all the List examples, we retrieve the list under the key `players` as if it was previously defined in [`GameServer.Spec.lists[players]`]({{< ref "/docs/Reference/agones_crd_api_reference.html#agones.dev/v1.GameServerSpec" >}}).

For your own List REST based requests, replace the value `players` with your own key in the path.

##### Alpha: GetList
This function retrieves a specific list identified by its name, returns the list's information.

###### Example
```bash
curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/lists/players
```

Response:
```json
{"name":"players", "capacity":"100", "values":["player0", "player1", "player2"]}     
```

##### Alpha: UpdateList

> This function updates the specified list's properties, such as its capacity and values, returns the updated list details. This will overwrite all existing List.Values with the update list request values. Use addValue or removeValue for modifying the List.Values field.

Reviewer: Can we please do `players` in this List section like we did for `rooms` above, and then I think this PR is good to go.

```suggestion
This function updates the list's properties with the key `players`, such as its capacity and values, returns the updated list details. This will overwrite all existing List.Values with the update list request values. Use addValue or removeValue for modifying the List.Values field.
```

###### Example

```bash
curl -d '{"capacity": "120", "values": ["player3", "player4"]}' -H "Content-Type: application/json" -X PATCH http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/lists/players
```

Response:
```json
{"name":"players", "capacity":"120", "values":["player3", "player4"]}
```

##### Alpha: AddListValue

> This function adds a new value to a specified list and returns the list with this addition.

Reviewer: ```suggestion
This function adds a new value to a list with the key `players` and returns the list with this addition.
```

###### Example     

```bash
curl -d '{"value": "player9"}' -H "Content-Type: application/json" -X POST http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/lists/players:addValue
```

Response:
```json
{"name":"players", "capacity":"120", "values":["player3", "player4", "player9"]}
```

##### Alpha: RemoveListValue

> This function removes a value from a given list and returns updated list.

Reviewer: ```suggestion
This function removes a value from the list with the key `players` and returns updated list.
```

curl -d '{"key": "foo", "value": "bar"}' -H "Content-Type: application/json" -X PUT http://localhost:${AGONES_SDK_HTTP_PORT}/metadata/annotation
```

> {{% feature publishVersion="1.38.0" %}}

Reviewer: No need to feature gate these docs either, since it's already been released 👍🏻

> {{% feature publishVersion="1.38.0" %}}

Reviewer 2: ```suggestion
```

### Counters and Lists

{{< alpha title="Counters and Lists" gate="CountsAndLists" >}}

> ### Counters

Reviewer: ```suggestion
#### Counters
```

Just so it's semantically under "Counters and Lists"

In all the  Counter examples, we retrieve the counter under the key `rooms` as if it was previously defined in [`GameServer.Spec.counters[room]`]({{< ref "/docs/Reference/agones_crd_api_reference.html#agones.dev/v1.GameServerSpec" >}}).

For your own Counter REST requests, replace the value `rooms` with your own key in the path.

> #### Alpha: GetCounter

Reviewer: ```suggestion
##### Alpha: GetCounter
```

Same for the other headers under `#### Counters`

This function retrieves a specified counter by its name and returns its information.

##### Example

```bash
curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/rooms
```

Response:
```json
{"name":"rooms", "count":"1", "capacity":"10"}
```

#### Alpha: UpdateCounter
This function updates the specified counter's properties, such as its count and capacity, and returns the updated counter details.

##### Example

```bash
curl -d '{"count": "5", "capacity": "11"}' -H "Content-Type: application/json" -X PATCH http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/rooms
```

Response:
```json
{"name":"rooms", "count":"5", "capacity":"11"}
```

> ### Lists

Reviewer: ```suggestion
#### Lists
```

In all the List examples, we retrieve the list under the key `players` as if it was previously defined in [`GameServer.Spec.lists[players]`]({{< ref "/docs/Reference/agones_crd_api_reference.html#agones.dev/v1.GameServerSpec" >}}).

For your own List REST based requests, replace the value `players` with your own key in the path.

> #### Alpha: GetList

Reviewer: ```suggestion
##### Alpha: GetList
```

and again, same as below.

curl -d '{"key": "foo", "value": "bar"}' -H "Content-Type: application/json" -X PUT http://localhost:${AGONES_SDK_HTTP_PORT}/metadata/annotation
```

> ### Local Development Setup: Counters and Lists

Reviewer: ```suggestion
### Counters and Lists
```

Author: I haven't included it. Is it really necessary? 🤔

> Before using the examples provided in this document for Counters and Lists, please ensure that you have set up the local development version with `CountsAndLists` enabled. You can do this by running the following command:
> 
> ```bash
> FEATURE_GATES="CountsAndLists=true" go run cmd/sdk-server/main.go --local
> ```
> 
> This command starts the local development server, which includes a Counter named `conformanceTestCounter` and a List named `conformanceTestList`.
> 
> Additionally, please make sure to set the HTTP port to 9358 as the local server uses it. You can do this by running the following command in a different terminal:
> 
> ```bash
> export AGONES_SDK_HTTP_PORT=9358
> ```

Reviewer: ```suggestion
```

Delete all this please - this page isn't covering the local SDK usage, just the REST API.

Reviewer: As a note - we could make a section in https://agones.dev/site/docs/guides/client-sdks/local/ on how to enable feature flags (which we have above), which would be useful! @igooch wdyt?

Author: I've added sections for `Enabling Specific Features` and `Setting the HTTP port`.

curl -d '{"key": "foo", "value": "bar"}' -H "Content-Type: application/json" -X PUT http://localhost:${AGONES_SDK_HTTP_PORT}/metadata/annotation
```

### Counters

{{< alpha title="Counters" gate="Counters" >}}

#### Alpha: GetCounter
This function retrieves a specified counter by its name and returns its information.

##### Example

```bash

> curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/conformanceTestCounter

Reviewer 2: We should probably mention in the example that conformanceTestCounter is predefined in the local SDK Server, or go with the more generic example of {name} or {counterName}.

Author: I have included the Local Development Setup section for Counters and Lists in this document. Could you please review it and share your thoughts?

Reviewer: I would add an explainer under each example that uses a key, something like:

> In this example, we retrieve the counter under the key "conformanceTestCounter" to make it extra clear.

Since people don't always read the top level docs 😃

As an improvement to the docs and likely the local sdk experience,  @igooch I didn't catch this earlier, but any objection to changing:
https://github.com/googleforgames/agones/blob/1e78d88a27ae7d427e43f073aab7d28289b7d850/pkg/sdkserver/localsdk.go#L147-L158

To using some more generic keys? Maybe `players` for the list and `rooms` for the count? 

We don't actually use these values for conformance tests anyway, so this might be a bit confusing, and I think would make examples easier to understand. WDYT? If so, we can do this in a separate PR (and will need to feature shortcode the new docs) - if you like we can file a new issue.

Author: >I would add an explainer under each example that uses a key, something like:
In this example, we retrieve the counter under the key "conformanceTestCounter" to make it extra clear.

I have included this information in a `Note` section located under both the `Counters` and `Lists` titles, following the `Warning` section. This will avoid the repetition right? WDYT?

>To using some more generic keys? Maybe players for the list and rooms for the count?

Do you want to replace `"conformanceTestCounter": {Count: 1, Capacity: 10}} ` with `"rooms": {Count: 1, Capacity: 10}} ` and `"conformanceTestList": {Values: []string{"test0", "test1", "test2"}, Capacity: 100}}` with `"players": {Values: []string{"player1", "player2", "player3"}, Capacity: 100}}`?

Reviewer: > We don't actually use these values for conformance tests anyway

I'm silly - of course we use this for conformance tests 🤦🏻 . Ignore that comment. But I am wondering if we change the key anyway, to make it easier for end users, and make our conformance testing match. WDYT?

For these docs, we don't have to conform to the local SDK for the example, so I think we can change the key to something more generic like I suggest above. I'd like to get agreement on the code changes from @igooch before we make the code changes though to the local sdk server (and either way, we should handle that in a separate PR).

Sound good?

Reviewer 2: My personal preference is to have the example match what we have in the localSDKServer for a better user experience. That way the user can run the example code and it works without them needing to make changes. My general gripe with generic examples is that they implicitly assume the user knows what and where to make changes without any accompanying instructions. That being said the key "conformanceTestCounter" probably doesn't mean anything to the user either, so I'd rename if along the lines of "TestCounterKey", "TestPlayerCounter", etc.

I'm also fine leaving the examples more generic here, as long as we have a step by step guide elsewhere that is more accessible to the user that needs more context on how and where to make changes.

Reviewer: > My personal preference is to have the example match what we have in the localSDKServer for a better user experience.

100% agreed - no disagreement there - sounds like we're just talking about what names to give keys. This change I've suggested above I would like to see as an interim step until we can align it again next release with a future PR to update the local sdk server and conformance tests.

> That being said the key "conformanceTestCounter" probably doesn't mean anything to the user either, so I'd rename the keys along the lines of "TestCounterKey", "TestPlayerCounter", etc.

So i would argue that both "test", "key", and "counter" are redundant in the name, and just confuse things a bit more, and it's easier to give a real world use case for the local SDK user that is easy to understand (and will align with the eventual integration pattern docs I'll write - primarily one for "rooms" and one for "player tracking" - as well as the examples we have in the yaml files etc.).

That being said, I would be super happy if as part of the description for each method, there is a reminder for what the key is that is being used, since reminders are great and redundancy in docs is definitely not a bad thing, I'll write up a small suggestion, lemme know what you think.

Author: Are we sticking with `rooms` and `players` for now and considering changes later if needed?

```

Response:
```json
{"name":"conformanceTestCounter", "count":"1", "capacity":"10"}
```

#### Alpha: UpdateCounter
This function updates the specified counter's properties, such as its count and capacity, and returns the updated counter details.

##### Example

```bash
curl -d '{"count": "5", "capacity": "11"}' -H "Content-Type: application/json" -X PATCH http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/counters/conformanceTestCounter
```

Response:
```json
{"name":"conformanceTestCounter", "count":"5", "capacity":"11"}
```

### Lists

{{< alpha title="Lists" gate="Lists" >}}

#### Alpha: GetList
This function retrieves a specific list identified by its name, returns the list's information.

##### Example
```bash
curl -H "Content-Type: application/json" -X GET http://localhost:${AGONES_SDK_HTTP_PORT}/v1alpha1/lists/conformanceTestList
```

Response:
```json
{"name":"conformanceTestList", "capacity":"100", "values":["test0", "test1", "test2"]     
```

#### Alpha: UpdateList

> This function updates the specified list's properties, such as its capacity and values, returns the updated list details.

Reviewer 2: ```suggestion
This function updates the specified list's properties, such as its capacity and values, returns the updated list details. This will overwrite all existing List.Values with the update list request values. Use addValue or removeValue for modifying the List.Values field.
```
